embulk-util-throttle-helper/src/main/java/org/embulk/util/throttle_helper/QuotaRateLimiter.java

```diff
@@ -0,0 +1,157 @@
+package org.embulk.util.throttle_helper;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.util.concurrent.Uninterruptibles;
+
+import javax.annotation.concurrent.ThreadSafe;
+
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+
+/**
+ * A rate limiter that allows a specific number of operations happen within
+ * a time window and its rate limit is only reset at the beginning of the
+ * next time window.
+ * <p>
+ * The time window is currently in second resolution to make sure unused permit
+ * doesn't get lost in situation when a thread, that must sleep in more than 2
+ * time windows, wakes up before all other threads, and reset the limits.
+ *
+ * @author instcode
+ * @author thangnc
+ */
+@ThreadSafe
+public class QuotaRateLimiter
+{
+    /**
+     * This is to prevent multiple threads from updating {@link #remainingLimit} &
+     * {@link #lastRefreshTime} wrongly
+     */
+    private Lock lock = new ReentrantLock();
+
+    /**
+     * The number of permits per the given time window
+     */
+    private int rateLimit;
+
+    /**
+     * The amount of time that affects the rate limit, i.e. the number of
+     * operations should not exceed the rate limit within this time window.
+     * This is also the amount of time that the rate limit gets refreshed.
+     */
+    private int timeWindowInMillis;
+
+    /**
+     * The number of available permits to give. If it is a positive number, it's
+     * the number of remaining permits for the current time window. Otherwise, it
+     * indicates that there are pending reserved permits that may affect the time to
+     * wait of incoming requests. For example, assume that the rate limit is 5 in
+     * 10 seconds as starting point, then if there are 15 requests come at once,
+     * there will be at least 5 requesters should have to wait for 20 seconds and
+     * the remaining limit is -5
```

Comment:
One overall question. Do you think is it possible that a thread can be in a "starvation" depending on the order where threads wake up from sleep, or it never happens?

The code looks that multiple threads wake up almost the same time. (Correct me if the understanding is wrong.) If so, if a thread No.1 always wakes up earlier than other threads, I'm afraid that other threads may be in starvation.

Reply:
I think thread starving shouldn't be the case because the waiting time for each thread is calculated based on the number of permits requested & remaining (negative number means more time to wait).

Reply:
Okay, just wanted to confirm.